scarif/rail_control.py

```diff
@@ -0,0 +1,403 @@
+    timeout: integer : USB serial communcations timeout in seconds
+
+    Raises
+    ------
+    GrblSerialException: The serial port is not open
+    GrblCommandException: Basic command checking before sending
+    """
+    def __init__(self, port, baudrate, timeout):
+        super(GrblCtrl, self).__init__()
+        self._serial = Serial()
+        self._serial.port = port
+        self._serial.baudrate = baudrate
+        self._serial.timeout = timeout
+
+    def open(self):
+        self._serial.open()
+
+    def close(self):
+        self._serial.close()
+
+    def _command(self, command):
+        """Wrapper for executing a grbl command"""
+        if not self._serial.isOpen():
+            raise GrblSerialException('%s port is not open' % (self.port))
+        result = self._serial.write(command+'\n')
+        logger.debug('cmd: "%s" ret %i' % (command, result))
+        return self._response()
+
+    def _response(self):
+        """Wrapper for grbl command response"""
+        result = []
+        logger.debug('res:')
+        done = False
+        while not done:
+            raw_result = self._serial.readline()
+            logger.debug(repr(raw_result))
+            if raw_result == '':
+                # TODO: Might be better to raise GrblTimeoutException here,
+                #       something like:
+                # raise GrblTimeoutException(
+                #     'Timeout encountered on read.')
+                logger.warning('Timeout encountered on read.')
+                done = True
+            if raw_result == 'ok\r\n':
+                done = True
+            result.append(raw_result.strip())
+        return result
+
+    # The '$'-commands are Grbl system commands used to tweak the settings,
+    # view or change Grbl's states and running modes, and start a homing cycle.
+
+    def _system_command(self, x):
+        if not x.startswith('$'):
+            raise GrblCommandException(
+                '%s does not start with $' % (x))
+        return self._command(x)
+
+    def _g_command(self, x):
+        if not x.startswith('G'):
+            raise GrblCommandException(
+                '%s does not start with G' % (x))
+        return self._command(x)
+
+    def view_grbl_settings(self):
+        """$$ (view Grbl settings)"""
+        return self._system_command('$$')
+
+    def view_hash_parameters(self):
+        """$# (view # parameters)"""
+        return self._system_command('$#')
+
+    def view_parser_state(self):
+        """$G (view parser state)"""
+        return self._system_command('$G')
+
+    def view_build_info(self):
+        """$I (view build info)"""
+        return self._system_command('$I')
+
+    def view_startup_blocks(self):
+        """$N (view startup blocks)"""
+        return self._system_command('$N')
+
+    def save_grbl_setting(self, x, value):
+        """$x=value (save Grbl setting)"""
+        return self._system_command('$%s=%s' % (str(x), str(value)))
```

Comment:
Not wrong but format is also an option.
e.g. self._system_command('${}={}'.format(x, value))

Reply:
Thanks. .format looks better.